core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -394,7 +394,8 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
     newProps ++= staticBrokerConfigs
     overrideProps(newProps, dynamicDefaultConfigs)
     overrideProps(newProps, dynamicBrokerConfigs)
-    val oldConfig = currentConfig
+    val oldConfig = new KafkaConfig(currentConfig.props, false, None)
+
```

Comment:
In current code,  oldconfig  refers to currentConfig /kafkaconfig reference. Few lines below while calling  "kafkaConfig.updateCurrentConfig(newConfig)". This will also update oldconfig. After this both oldconfigm and newconfig will point to same config

Reply:
@omkreddy Did you run into any issues with this? I think the only time `kafkaConfig` is the same as `currentConfig` is during initialization when we are not relying on `reconfigure` and the values in `oldConfig` are not used. Thereafter, `kafkaConfig.updateCurrentConfig(newConfig)` shouldn't impact `oldConfig`. Have I missed something?

Reply:
@rajinisivaram Yes, while updating Broker configs like no.io.threads, listeners etc, i was getting same values for both old and new configs.  I added few SOPs to make sure they are same.

In current code, below are the sequence of assignments. I think "oldConfig" still refering kafkaConfig

```
private var currentConfig = kafkaConfig
val oldConfig = currentConfig
currentConfig = newConfig
kafkaConfig.updateCurrentConfig(newConfig)
```

With above change, I was able to update the configs.

Reply:
When `DynamicBrokerConfig.initialize()` is invoked, `currentConfig` is updated using `updateCurrentConfig`. At this point, `currentConfig` is `kafkaConfig`, but there are no reconfigurables that need to be reconfigured, so `oldConfig` is not used. Subsequent invocations of `updateCurrentConfig` should have `currentConfig` that is not the same as `kafkaConfig`. When you were running into this issue, were you testing configuration update of a server or a broker restart or a unit test?

Reply:
@rajinisivaram  I was testing config update of a server

Sample steps to reproduce:
1.   start a single node cluster
2.  Try to update num.io.threads (default : 8)
sh bin/kafka-configs.sh --bootstrap-server localhost:9092 --entity-type brokers --entity-name 0 --alter --add-config num.io.threads=10
3.   No changes to  num.io.threads ( jstack shows 8 threads)

with my patch, I will get below message on server side and jstack shows 10 threads.
```
[Kafka Request Handler on Broker 0], Resizing request handler thread pool size from 8 to 10
```

Reply:
@omkreddy Ah, `kafkaConfig` doesn't get updated during `initialize` if there are no dynamic configs defined in ZK. I think a better fix would be to change the initialization:
`private var currentConfig = new KafkaConfig(kafkaConfig.props, false, None)`.

Reply:
Above line will go into infinite loop.
```
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:992)
	at kafka.server.DynamicBrokerConfig.<init>(DynamicBrokerConfig.scala:136)
	at kafka.server.KafkaConfig$$anonfun$3.apply(KafkaConfig.scala:992)
	at kafka.server.KafkaConfig$$anonfun$3.apply(KafkaConfig.scala:992)
	at scala.Option.getOrElse(Option.scala:121)
	at kafka.server.KafkaConfig.<init>(KafkaConfig.scala:992)
	at kafka.server.DynamicBrokerConfig.<init>(DynamicBrokerConfig.scala:136)
	at kafka.server.KafkaConfig$$anonfun$3.apply(KafkaConfig.scala:992)
```
Let me know if any changes required. 
Raised JIRA: https://issues.apache.org/jira/browse/KAFKA-6680
Do you think, this should be part of 1.1 release?

Reply:
@omkreddy Sorry, we should update `currentConfig` in `DynamicBrokerConfig.initialize`. Thanks for raising the JIRA. While it would have been good to have this in 1.1, I am thinking we don't need to make this a blocker since it only impacts the first time a dynamic config is added. We will need a restart the first time and thereafter it should work as expected. We could highlight this in the docs for 1.1. What do you think?

Reply:
Actually, a restart is not required, since a second update would work, so the config can be removed and added again (not ideal, but perhaps will do for 1.1).

Reply:
@rajinisivaram Updated the PR. Pls take a look. Agree, we can document this behaviour for 1.1